examples/colors_rgb.rs

```diff
@@ -141,75 +150,113 @@ impl Default for Fps {
-                colors: &app.colors,
-                frame_count: app.frame_count,
-            },
+/// Widget impl for FpsWidget
+///
+/// This is implemented on a mutable reference so that we can update the frame count and fps
+/// calculation while rendering.
+impl Widget for &mut FpsWidget {
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        self.calculate_fps();
+        if let Some(fps) = self.fps {
+            let text = format!("{:.1} fps", fps);
+            Text::from(text).render(area, buf);
         }
     }
 }
 
-impl Widget for AppWidget<'_> {
-    fn render(self, area: Rect, buf: &mut Buffer) {
-        let vertical = Layout::vertical([Constraint::Length(1), Constraint::Min(0)]);
-        let horizontal = Layout::horizontal([Constraint::Min(0), Constraint::Length(8)]);
-        let [top, colors] = area.split(&vertical);
-        let [title, fps] = top.split(&horizontal);
-
-        self.title.render(title, buf);
-        self.fps_widget.render(fps, buf);
-        self.rgb_colors_widget.render(colors, buf);
+impl FpsWidget {
+    /// Update the fps calculation.
+    ///
+    /// This updates the fps once a second, but only if the widget has rendered at least 2 frames
+    /// since the last calculation. This avoids noise in the fps calculation when rendering on slow
+    /// machines that can't render at least 2 frames per second.
+    fn calculate_fps(&mut self) {
+        self.frame_count += 1;
+        let elapsed = self.last_instant.elapsed();
+        if elapsed > Duration::from_secs(1) && self.frame_count > 2 {
+            self.fps = Some(self.frame_count as f32 / elapsed.as_secs_f32());
+            self.frame_count = 0;
+            self.last_instant = Instant::now();
+        }
     }
 }
 
-impl Widget for RgbColorsWidget<'_> {
+/// Widget impl for ColorsWidget
+///
+/// This is implemented on a mutable reference so that we can update the frame count and store a
+/// cached version of the colors to render instead of recalculating them every frame.
+impl Widget for &mut ColorsWidget {
+    /// Render the widget
     fn render(self, area: Rect, buf: &mut Buffer) {
-        let colors = self.colors;
+        self.setup_colors(area);
+        let colors = &self.colors;
         for (xi, x) in (area.left()..area.right()).enumerate() {
             // animate the colors by shifting the x index by the frame number
             let xi = (xi + self.frame_count) % (area.width as usize);
             for (yi, y) in (area.top()..area.bottom()).enumerate() {
+                // render a half block character for each row of pixels with the foreground color
+                // set to the color of the pixel and the background color set to the color of the
+                // pixel below it
                 let fg = colors[yi * 2][xi];
                 let bg = colors[yi * 2 + 1][xi];
                 buf.get_mut(x, y).set_char('▀').set_fg(fg).set_bg(bg);
             }
         }
+        self.frame_count += 1;
```

Comment:
We could use the `frame_count` in frame, that would make an example usage.

Reply:
The problem is that it's not available from the widget.render method as you don't have access to frame there. So it would have to be pushed down to each widget as part of the main draw call, which makes this a little messier than I'd like. Not sure if there's a simple solution to that that I've missed.

Reply:
Oh right, that's fine then